### PR TITLE
Accept application/x-www-form-urlencoded webhooks

### PR DESCRIPTION
**Pull request by lkysow**
- Enable `ValidatePayload` to work for webhooks with Content-Type `application/x-www-form-urlencoded` in addition to `application/json`.
- Reject any Content-Type that is not `application/x-www-form-urlencoded` or `application/json`.

### CODE REVIEW

File: github/messages_test.go [hunk 1]
```diff
@@ -82,6 +83,54 @@ func TestValidatePayload(t *testing.T) {
 	}
 }
 
+func TestValidateFormPayloadGET(t *testing.T) {
```
Comment (lkysow) [inline github/messages_test.go]: I looked at adding this test to the original `TestValidatePayload` test cases but thought the additional complexity of adding it to that test was too much since you need to change how the test is constructing the request.

**Comment by lkysow** [comment]
I signed it!

File: github/messages.go [hunk 2]
```diff
@@ -132,17 +137,37 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
+	if r.Header.Get("Content-Type") == "application/x-www-form-urlencoded" {
+		// If the content type is application/x-www-form-urlencoded
+		// the json payload will be under the "payload" form param
+		jsonPayload = []byte(r.FormValue(payloadFormParam))
+
+		// GitHub calculates the signature based on the query-escaped
+		// post body. In order to validate, we need to reconstruct the raw body.
+		rawPayload = []byte(fmt.Sprintf("%s=%s", payloadFormParam, url.QueryEscape(string(jsonPayload))))
+	} else {
+		defer r.Body.Close()
```
Comment (shurcooL) [inline github/messages.go]: `r.Body.Close()` wasn't being done before. Why is it done now?

File: github/messages.go [hunk 3]
```diff
@@ -122,6 +126,7 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
 
 // ValidatePayload validates an incoming GitHub Webhook event request
 // and returns the (JSON) payload.
+// The payload can be of Content-Type application/json or application/x-www-form-urlencoded
```
Comment (shurcooL) [inline github/messages.go]: Nitpicks.

This sentence is missing a period at the end.

Also, I'd rephrase it as:

```
// The payload can have content type of "application/json" or "application/x-www-form-urlencoded".
```

`Content-Type` is the name of the header, but it's pretty natural to use the phrase "content type" or "type of content" in English. E.g., [`ProjectCardOptions.ContentType`](https://godoc.org/github.com/google/go-github/github#ProjectCardOptions.ContentType) is documented as "The type of content to associate with this card."

File: github/messages_test.go (hunk 1 above)
Comment (shurcooL) [inline github/messages_test.go]: That's a good call. It's better to add a separate test, then we can be more confident the old one still passes without changes.

**Comment by lkysow** [comment]
@shurcooL regarding your question, the two different types of webhooks are documented here:
https://developer.github.com/webhooks/creating/#content-type

If you click "Add webhook" this is the screen you'll see where by default the formencoded Content-Type is selected. In our application that asks users to add webhooks we found that sometimes users would add the `formencoded` content-type (even though our docs say use json), so we wanted to support both webhook types.

image

**Comment by shurcooL** [comment]
Awesome, thanks, that answers my question. Since GitHub supports both types, it would be a good enhancement for go-github to support them too. Thanks for sending the PR. My next step will be to review the code, I will get to that in a bit.

File: github/messages.go (hunk 2 above)
Comment (lkysow) [inline github/messages.go]: I think it was a bug with the original implementation. If you look at the docs for `http`: https://golang.org/pkg/net/http/ it's in the example. If you don't close it then the connection remains open: https://stackoverflow.com/questions/33238518/what-could-happen-if-i-dont-close-response-body-in-golang

image

File: github/messages_test.go [hunk 4]
```diff
@@ -82,6 +83,54 @@ func TestValidatePayload(t *testing.T) {
...
+	secretKey := []byte("0123456789abcdef")
+
+	req, err := http.NewRequest("GET", "http://localhost/event", nil)
+	q := req.URL.Query()
+	q.Add("payload", payload)
+	req.URL.RawQuery = q.Encode()
+	if err != nil {
+		t.Fatalf("NewRequest: %v", err)
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
```
Comment (shurcooL) [inline github/messages_test.go]: Since you only ever want one value for Content-Type, use `req.Header.Set` rather than `req.Header.Add`. It doesn't make a difference as long as you do it only once, but still, using `Set` is better form and more readable and consistent code.

File: github/messages_test.go [hunk 5]
```diff
@@ -82,6 +83,54 @@ func TestValidatePayload(t *testing.T) {
...
+	secretKey := []byte("0123456789abcdef")
+
+	form := url.Values{}
+	form.Set("payload", payload)
+	buf := bytes.NewBufferString(form.Encode())
+	req, err := http.NewRequest("POST", "http://localhost/event", buf)
+	if err != nil {
+		t.Fatalf("NewRequest: %v", err)
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
```
Comment (shurcooL) [inline github/messages_test.go]: Use `req.Header.Set("Content-Type", ...)` here as well.

File: github/messages_test.go (hunk 1 above)
Comment (shurcooL) [inline github/messages_test.go]: Rename the test to `TestValidatePayload_FormGet` for more consistent test naming.

File: github/messages_test.go [hunk 6]
```diff
@@ -82,6 +83,54 @@ func TestValidatePayload(t *testing.T) {
...
+	got, err := ValidatePayload(req, secretKey)
+	if err != nil {
+		t.Errorf("ValidatePayload(%#v): err = %v, want nil", payload, err)
+	}
+	if string(got) != payload {
+		t.Errorf("ValidatePayload = %q, want %q", got, payload)
+	}
+}
+
+func TestValidateFormPayloadPOST(t *testing.T) {
```
Comment (shurcooL) [inline github/messages_test.go]: And `TestValidatePayload_FormPost` here.

File: github/messages.go [hunk 7]
```diff
@@ -132,17 +137,37 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+func ValidatePayload(r *http.Request, secretKey []byte) ([]byte, error) {
+	// jsonPayload is the json webhook data
+	var jsonPayload []byte
+	// rawPayload is the raw body that GitHub uses to calculate the signature
+	var rawPayload []byte
+
+	if r.Header.Get("Content-Type") == "application/x-www-form-urlencoded" {
```
Comment (shurcooL) [inline github/messages.go]: Since "application/x-www-form-urlencoded" and "application/json" are the only two content types we expect from GitHub, and support, let's explicitly check it's one of those two and return an error otherwise. It wouldn't make sense to treat it as JSON if the content type is not actually "application/json".

This is a good place to use a `switch`, like this:

```Go
switch r.Header.Get("Content-Type") {
case "application/x-www-form-urlencoded":
    ...
case "application/json":
    ...
default:
    return nil, fmt.Errorf("Webhook payload has unsupported Content-Type %q", r.Header.Get("Content-Type"))
}
```

File: github/messages_test.go [hunk 8]
```diff
@@ -82,6 +83,54 @@ func TestValidatePayload(t *testing.T) {
...
+
+	got, err := ValidatePayload(req, secretKey)
+	if err != nil {
+		t.Errorf("ValidatePayload(%#v): err = %v, want nil", payload, err)
+	}
+	if string(got) != payload {
+		t.Errorf("ValidatePayload = %q, want %q", got, payload)
+	}
+}
+
```
Comment (shurcooL) [inline github/messages_test.go]: Since the primary purpose of `ValidatePayload` is to reject invalid payloads, can you also add a test that for an invalid payload that correctly returns a non-nil error?

File: github/messages.go (hunk 7 above)
Comment (lkysow) [inline github/messages.go]: Totally agree. FYI though that this will be a backwards incompatible change since previously we never checked the content-type header and now we're rejecting anything without `application/json`. If people are using this as expected then GitHub would always send that header, however it is technically breaking BC.

**Comment by lkysow** [comment]
@shurcooL updated

File: github/messages.go (hunk 2 above)
Comment (shurcooL) [inline github/messages.go]: I see.

However, this function is meant to be used on in HTTP handlers working on server HTTP requests, not client ones.

Specifically:

```Go
func (s *GitHubEventMonitor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
    payload, err := github.ValidatePayload(r, s.webhookSecretKey)
    if err != nil { ... }
    // Process payload...
}
```

HTTP handlers do not need to close the response body.

See documentation here:

https://godoc.org/net/http#Request.Body

> For server requests the Request Body is always non-nil
> but will return EOF immediately when no body is present.
> **The Server will close the request body. The ServeHTTP
> Handler does not need to.**

So, the original behavior was correct, and you should revert that change.

File: github/messages.go (hunk 7 above)
Comment (shurcooL) [inline github/messages.go]: I think that's fine, because GitHub will always be setting that header.

File: github/messages.go [hunk 9]
```diff
@@ -132,17 +138,40 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
 //
-func ValidatePayload(r *http.Request, secretKey []byte) (payload []byte, err error) {
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+func ValidatePayload(r *http.Request, secretKey []byte) ([]byte, error) {
+	// jsonPayload is the json webhook data
+	var jsonPayload []byte
+	// rawPayload is the raw body that GitHub uses to calculate the signature
+	var rawPayload []byte
```
Comment (shurcooL) [inline github/messages.go]: Style wise, it might be nicer to do this:

```Go
var (
    jsonPayload []byte // JSON webhook data.
    rawPayload  []byte // Raw body that GitHub uses to calculate the signature.
)
```

But it's up to you, use what you think looks nicer.

File: github/messages.go [hunk 10]
```diff
@@ -132,17 +138,40 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
+		// post body. In order to validate, we need to reconstruct the raw body.
+		rawPayload = []byte(fmt.Sprintf("%s=%s", payloadFormParam, url.QueryEscape(string(jsonPayload))))
+	case "application/json":
+		defer r.Body.Close()
+		var err error
+		jsonPayload, err = ioutil.ReadAll(r.Body)
+		// If the content type is application/json then the raw body
+		// used to calculate the signature is just the original body.
+		rawPayload = jsonPayload
+		if err != nil {
```
Comment (shurcooL) [inline github/messages.go]: This `if err != nil` check is out of place. Did you mean to place it directly under `jsonPayload, err = ioutil.ReadAll(r.Body)` line?

File: github/messages.go [hunk 11]
```diff
@@ -132,17 +138,40 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
+
+	switch c := r.Header.Get("Content-Type"); c {
+	case "application/x-www-form-urlencoded":
+		// If the content type is application/x-www-form-urlencoded
+		// the json payload will be under the "payload" form param
+		jsonPayload = []byte(r.FormValue(payloadFormParam))
+
+		// GitHub calculates the signature based on the query-escaped
+		// post body. In order to validate, we need to reconstruct the raw body.
+		rawPayload = []byte(fmt.Sprintf("%s=%s", payloadFormParam, url.QueryEscape(string(jsonPayload))))
```
Comment (shurcooL) [inline github/messages.go]: It feels to me that using `fmt.Sprintf` here is not being helpful, and the code would be simpler with just plain string concatenation.

How about:

```Go
rawPayload = []byte(payloadFormParam + "=" + url.QueryEscape(string(jsonPayload)))
```

File: github/messages.go [hunk 12]
```diff
@@ -132,17 +138,40 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+func ValidatePayload(r *http.Request, secretKey []byte) ([]byte, error) {
+	// jsonPayload is the json webhook data
+	var jsonPayload []byte
+	// rawPayload is the raw body that GitHub uses to calculate the signature
+	var rawPayload []byte
+
+	switch c := r.Header.Get("Content-Type"); c {
```
Comment (shurcooL) [inline github/messages.go]: I think `ct` would be a better, more readable variable name for Content-Type.

File: github/messages.go (hunk 2 above)
Comment (lkysow) [inline github/messages.go]: :facepalm: nice catch, will do

File: github/messages.go (hunk 10 above)
Comment (lkysow) [inline github/messages.go]: yes, moving

File: github/messages.go (hunk 9 above)
Comment (lkysow) [inline github/messages.go]: I like that better too

**Comment by lkysow** [comment]
@shurcooL thanks for the review! Updated again based on comments.

**Comment by lkysow** [comment]
@shurcooL @gmlewis any chance you guys can look at this again? I've made all requested changed.

File: github/messages.go [hunk 13]
```diff
@@ -132,17 +138,39 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
+	)
+	switch ct := r.Header.Get("Content-Type"); ct {
+	case "application/x-www-form-urlencoded":
+		// If the content type is application/x-www-form-urlencoded
+		// the json payload will be under the "payload" form param.
+		jsonPayload = []byte(r.FormValue(payloadFormParam))
+
+		// GitHub calculates the signature based on the query-escaped
+		// post body. In order to validate, we need to reconstruct the raw body.
+		rawPayload = []byte(payloadFormParam + "=" + url.QueryEscape(string(jsonPayload)))
```
[the remark this answers is not in the document]
Comment (lkysow) [reply github/messages.go]: @shurcooL agreed this is more straight-forward. My only hesitation is around your suggestion to limit the POST body to 10MB. Wouldn't we have to copy the code from https://golang.org/src/net/http/request.go#L1101

```
		var reader io.Reader = r.Body
		maxFormSize := int64(1<<63 - 1)
		if _, ok := r.Body.(*maxBytesReader); !ok {
			maxFormSize = int64(10 << 20) // 10 MB is a lot of text.
			reader = io.LimitReader(r.Body, maxFormSize+1)
		}
		b, e := ioutil.ReadAll(reader)
		if e != nil {
			if err == nil {
				err = e
			}
			break
		}
		if int64(len(b)) > maxFormSize {
			err = errors.New("http: POST too large")
			return
		}
		vs, e = url.ParseQuery(string(b))
		if err == nil {
			err = e
		}
```

Then I'm just worried that we're re-implementing a lot of logic from the core.

[the remark this answers is not in the document]
Comment (shurcooL) [reply github/messages.go]: We don't have to copy all of that code, nor do we have to worry about keeping it in sync. We just need to consider putting in a similar 10 MB limit.

If we decide to do it, we can look at the net/http implementation for inspiration, but I think ours can be simpler:

```Go
const maxFormSize = int64(10 << 20)

var err error
body, err = ioutil.ReadAll(io.LimitReader(r.Body, maxFormSize+1))
if err != nil {
	return nil, err
}
if int64(len(body)) > maxFormSize {
	return nil, errors.New("POST too large")
}
```

Also, we should think about why there's a limit for POST but not for JSON. Does JSON need one too? Or is it okay not to have one for either?

[the remark this answers is not in the document]
Comment (lkysow) [reply github/messages.go]: I've pushed your suggested changes *without* the 10MB limits. I've tested this locally with/without the secret being specified and with both Content-Types and it works as expected.

RE: 10MB limit, I'm still not sure what the right answer is. Can we get this in and then cut another issue to track it as it seems like a separate concern? I'm happy to create that.

[the remark this answers is not in the document]
Comment (shurcooL) [reply github/messages.go]: I'm okay with making it a separate issue. This affects only the new feature, not the existing functionality anyway. I think we should gather information in that issue and only add the limit once we have a provable situation for when it's needed. Thanks.

**Comment by lkysow** [comment]
Sounds good to me 👍  thanks for your excellent code review and explaining why you request certain changes. Really appreciate it.

**Comment by lkysow** [comment]
Created #735 to track cap on body size

**Comment by lkysow** [comment]
@gmlewis 
>  I'll ask @gmlewis to confirm the latest changes are okay with him too.
from above

**Comment by gmlewis** [comment]
Sorry for the delay... taking a look now.

File: github/messages_test.go [hunk 14]
```diff
@@ -82,6 +85,80 @@ func TestValidatePayload(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("ValidatePayload(%#v): err = %v, want nil", payload, err)
+	}
+	if string(got) != payload {
+		t.Errorf("ValidatePayload = %q, want %q", got, payload)
+	}
+
+	// check that if payload is invalid we get error
+	req.Header.Set(signatureHeader, "invalid signature")
+	_, err = ValidatePayload(req, nil)
```
Comment (gmlewis) [inline github/messages_test.go]: nit:

```go
if _, err = ValidatePayload(req, nil); err == nil {
```

for Go style.

File: github/messages_test.go [hunk 15]
```diff
@@ -82,6 +85,80 @@ func TestValidatePayload(t *testing.T) {
...
```
Comment (gmlewis) [inline github/messages_test.go]: nit - same here..

```go
if _, err = ValidatePayload(req, nil); err == nil {
```

File: github/messages_test.go [hunk 16]
```diff
@@ -82,6 +85,80 @@ func TestValidatePayload(t *testing.T) {
...
+	}
+}
+
+func TestValidatePayload_InvalidContentType(t *testing.T) {
+	req, err := http.NewRequest("POST", "http://localhost/event", nil)
+	if err != nil {
+		t.Fatalf("NewRequest: %v", err)
+	}
+	req.Header.Set("Content-Type", "invalid content type")
+	_, err = ValidatePayload(req, nil)
```
Comment (gmlewis) [inline github/messages_test.go]: nit - and here...

```go
if _, err = ValidatePayload(req, nil); err == nil {
```

File: github/messages.go [hunk 17]
```diff
@@ -133,13 +136,45 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
+		// the JSON payload is just the original body.
+		payload = body
+
+	case "application/x-www-form-urlencoded":
+		// payloadFormParam is the name of the form parameter that the JSON payload
+		// will be in if a webhook has its content type set to application/x-www-form-urlencoded.
+		const payloadFormParam = "payload"
+
+		var err error
+		body, err = ioutil.ReadAll(r.Body)
```
Comment (gmlewis) [inline github/messages.go]: nit:

```go
if body, err = ioutil.ReadAll(r.Body); err != nil {
```

File: github/messages.go [hunk 18]
```diff
@@ -133,13 +136,45 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
...
 func ValidatePayload(r *http.Request, secretKey []byte) (payload []byte, err error) {
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+	var body []byte // Raw body that GitHub uses to calculate the signature.
+
+	switch ct := r.Header.Get("Content-Type"); ct {
+	case "application/json":
+		var err error
+		body, err = ioutil.ReadAll(r.Body)
```
Comment (gmlewis) [inline github/messages.go]: nit:

```go
if body, err = ioutil.ReadAll(r.Body); err != nil {
```

**Comment by gmlewis** [comment]
Please ignore the [googletbot comment](https://github.com/google/go-github/pull/693#issuecomment-332898713) as I have verified that the CLAs all LGTM.

**Comment by lkysow** [comment]
@gmlewis done!

File: github/messages.go (hunk 18 above)
Comment (gmlewis) [reply github/messages.go]: Oh, OK.  I thought it was more idiomatic Go, but that's OK.  Feel free to switch it back... I don't feel that strongly about it.  Sorry for the switch... I forgot you preferred the other way.

Comment (shurcooL) [reply github/messages.go]: > I thought it was more idiomatic Go

As far as I know, both are valid idiomatic Go. Which one is better to use depends on the context/situation.

I'll leave it as is, I don't feel very strongly either.
